### ISSUE TITLE
Program slow to simplify

### ISSUE THREAD

**Issue by zaxtax** [hakaru-dev/hakaru#21]
The following program is slow to simplify

```
x5 = x6 = fn x11 pair(prob, prob):
           (1 *
            1 *
            (match x11:
              (x27, x28):
               x29 = prob2real(x27)
               (match (not((x29 < 3)) && not((8 < x29))):
                 true:
                  (recip(real2prob((8 + negate(3)))) *
                   x30 = prob2real(x28)
                   (match (not((x30 < 1)) && not((4 < x30))):
                     true:
                      (recip(real2prob((4 + negate(1)))) *
                       x19_ = prob2real(x27)
                       integrate x32 from negate(prob2real(∞)) to prob2real(∞):
                        (exp((negate(((x32 + negate(nat2real(0))) ^ 2)) *
                              recip(prob2real((2 * (real2prob(x19_) ^ 2)))))) *
                         recip(real2prob(x19_)) *
                         recip(natroot((2 * pi), 2)) *
                         x21_ = prob2real(x28)
                         x22_ = real2prob(x21_)
                         ((exp((negate(((nat2real(17) + negate(x32)) ^ 2)) *
                                recip(prob2real((nat2prob(2) * (x22_ ^ 2)))))) *
                           recip(x22_)) *
                          recip(natroot((nat2prob(2) * pi), 2)) *
                          integrate x35 from negate(prob2real(∞)) to prob2real(∞):
                           (exp((negate(((x35 + negate(x32)) ^ 2)) *
                                 recip(prob2real((2 * (real2prob(x19_) ^ 2)))))) *
                            recip(real2prob(x19_)) *
                            recip(natroot((2 * pi), 2)) *
                            x24_ = real2prob(x21_)
                            ((exp((negate(((nat2real(13) + negate(x35)) ^ 2)) *
                                   recip(prob2real((nat2prob(2) * (x24_ ^ 2)))))) *
                              recip(x24_)) *
                             recip(natroot((nat2prob(2) * pi), 2)) *
                             1)))))
                     false: 0))
                 false: 0)
              _: 0))
     fn x5 pair(prob, prob):
      x0 <~ (match x5:
              (noiseT, noiseE):
               weight(1/2,
                      noiseT_ <~ uniform(nat2real(3), nat2real(8))
                      return (real2prob(noiseT_), noiseE)) <|>
               weight(1/2,
                      noiseE_ <~ uniform(nat2real(1), nat2real(4))
                      return (noiseT, real2prob(noiseE_))))
      return (x0, (x6(x0) * recip(x6(x5))))
fn x4 pair(prob, prob):
 x3 <~ x5(x4)
 (match x3:
   (x1, x2):
    x0 <~ weight(min(1, x2), return true) <|>
          weight(real2prob((prob2real(1) + negate(prob2real(min(1, x2))))),
                 return false)
    return (match x0:
             true: x1
             false: x4))
```

In particular it takes about 20 seconds

```
[zv@t420 hakaru (master)]% time simplify slow.hk > /dev/null                                                                                                                                                                                                            
simplify slow.hk > /dev/null  19.53s user 5.78s system 66% cpu 37.789 total
```

**Comment by JacquesCarette** [comment]
@yuriy0 Can you take a look at this?  It isn't as important as it once was, but it would still be good to know where the time seems to be going.

**Comment by ccshan** [comment]
The replacement of `int` by `Int` should probably happen in `haskell/Language/Hakaru/Pretty/Maple.hs` line 188, by analogy with lines 199 and 210.

**Comment by JacquesCarette** [comment]
Yes, we came to that same conclusion in a discussion that happened at about the same time you posted that comment.

What we can also do is to try to evaluate those integrals 'cleverly' (meaning: in context) rather than eagerly, which is what is currently done, and does not work.

**Comment by yuriy0** [comment]
Fixed by 5118851aa and 80d3966d.
